src/coreclr/nativeaot/Runtime/eventtrace.cpp

```diff
@@ -94,6 +94,67 @@ enum CallbackProviderIndex
     DotNETRuntimePrivate = 3
 };
 
+#ifdef FEATURE_ETW
+// EventFilterType identifies the filter type used by the PEVENT_FILTER_DESCRIPTOR
+enum EventFilterType
+{
+    // data should be pairs of UTF8 null terminated strings all concatenated together.
+    // The first element of the pair is the key and the 2nd is the value. We expect one of the
+    // keys to be the string "GCSeqNumber" and the value to be a number encoded as text.
+    // This is the standard way EventPipe encodes filter values
+    StringKeyValueEncoding = 0,
+    // data should be an 8 byte binary LONGLONG value
+    // this is the historic encoding defined by .NET Framework for use with ETW
+    LongBinaryClientSequenceNumber = 1
+};
+
+void ParseFilterDataClientSequenceNumber(
+    EVENT_FILTER_DESCRIPTOR * FilterData,
+    LONGLONG * pClientSequenceNumber)
+{
+    if (FilterData == NULL)
+        return;
+
+    if (FilterData->Type == LongBinaryClientSequenceNumber && FilterData->Size == sizeof(LONGLONG))
+    {
+        *pClientSequenceNumber = *(LONGLONG *) (FilterData->Ptr);
+    }
+    else if (FilterData->Type == StringKeyValueEncoding)
+    {
+        const char* buffer = reinterpret_cast<const char*>(FilterData->Ptr);
+        const char* buffer_end = buffer + FilterData->Size;
+
+        while (buffer < buffer_end)
+        {
+            const char* key = buffer;
+            size_t key_len = strnlen(key, buffer_end - buffer);
+            buffer += key_len + 1;
+
+            if (buffer >= buffer_end)
+                break;
+
+            const char* value = buffer;
+            size_t value_len = strnlen(value, buffer_end - buffer);
+            buffer += value_len + 1;
+
+            if (buffer >= buffer_end)
```

Comment:
I think this should be '>' no? Maybe I have the wrong mental model but this is what I was expecting:
FilterData->Ptr = 0x1000000, pointing to the UTF8 string "GCSeqNumber\025\0"
FilterData->Size = 15
key = 0x1000000
key_len = 11
value = 0x100000c (+12)
value_len = 2
buffer = 0x100000f (+15)
buffer_end = 0x100000f (+15)

Reply:
Ah, you're right, I tricked myself thinking that the break conditions would be the same.